### PR TITLE
fix(nn): disable Transformer sparsity fast path when attn_mask is present (fixes #166166)

### LINKED ISSUE #166166: MultiheadAttention fast_path is substantially less efficient than the alternative
### 🐛 Describe the bug

When using an attention mask in inference mode, the fast_path in MultiheadAttention is substantially less efficient than the "default" path, both VRAM-wise and time-wise. In my case, the default path uses 1GB VRAM, while the fast path uses 4GB VRAM.

Minimal example:

```
import torch
import torch.nn as nn
import gc

def test_transformer_memory(mode: str):
    """Test memory usage of transformer with attention mask."""
    if mode == "with":
        torch.backends.mha.set_fastpath_enabled(True)
    else:
        torch.backends.mha.set_fastpath_enabled(False)
    device = "cuda"
    torch.cuda.empty_cache()
    torch.cuda.reset_peak_memory_stats()
    gc.collect()
    
    # Create a small transformer
    encoder_layer = nn.TransformerEncoderLayer(
        d_model=256,
        nhead=4,
        dim_feedforward=512,
        batch_first=True,
    )
    model = nn.TransformerEncoder(encoder_layer, num_layers=6).to(device).eval()
    
    batch_size = 32
    seq_len = 1200
    
    # Create inputs
    src = torch.randn(batch_size, seq_len, 256, device=device)
    src_key_padding_mask = torch.zeros(batch_size, seq_len, device=device, dtype=torch.bool)
    
    # Create dummy attention mask
    attn_mask = torch.zeros(seq_len, seq_len, device=device, dtype=torch.bool)
    
    # Warmup
    print(f"Mode: {mode}")
    with torch.no_grad():
        _ = model(src, mask=attn_mask, src_key_padding_mask=src_key_padding_mask)
    
    torch.cuda.synchronize()
    peak = torch.cuda.max_memory_allocated() / 1024**2
    print(f"Peak memory: {peak:.2f} MB")
    
    del model
    torch.cuda.empty_cache()
    gc.collect()
    
    return peak

if __name__ == "__main__":
    print("Testing transformer memory with attention mask\n")
    with_peak = test_transformer_memory("with")
    without_peak = test_transformer_memory("without")
    
    print(f"\n{'='*50}")
    print(f"With fast path: {with_peak:.2f} MB")
    print(f"Without fast path: {without_peak:.2f} MB")
    print(f"Difference: {with_peak - without_peak:+.2f} MB ({((with_peak/without_peak - 1) * 100):+.1f}%)")
```

### Versions

Collecting environment information...
PyTorch version: 2.8.0a0+5228986c39.nv25.06
Is debug build: False
CUDA used to build PyTorch: 12.9
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version: 3.12.3 (main, Aug 14 2025, 17:47:21) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-160-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 565.57.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.10.2
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i9-14900K
CPU family:                              6
Model:                                   183
Thread(s) per core:                      2
Core(s) per socket:                      24
Socket(s):                               1
Stepping:                                1
CPU(s) scaling MHz:                      23%
CPU max MHz:                             6000.0000
CPU min MHz:                             800.0000
BogoMIPS:                                6374.40
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               896 KiB (24 instances)
L1i cache:                               1.3 MiB (24 instances)
L2 cache:                                32 MiB (12 instances)
L3 cache:                                36 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-31
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Mitigation; Clear Register File
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] intel-openmp==2021.4.0
[pip3] mkl==2021.1.1
[pip3] mkl-devel==2021.1.1
[pip3] mkl-include==2021.1.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] nvidia-cudnn-frontend==1.12.0
[pip3] nvtx==0.2.11
[pip3] onnx==1.17.0
[pip3] optree==0.16.0
[pip3] pynvjitlink==0.3.0
[pip3] pytorch-lightning==2.4.0
[pip3] pytorch-triton==3.3.0+git96316ce52.nvinternal
[pip3] tbb==2021.13.1
[pip3] torch==2.8.0a0+5228986c39.nv25.6
[pip3] torch_cluster==1.6.3
[pip3] torch_geometric==2.5.0
[pip3] torch_scatter==2.1.2
[pip3] torch_sparse==0.6.18
[pip3] torch_tensorrt==2.8.0a0
[pip3] torchao==0.11.0+git
[pip3] torchmetrics==1.8.2
[pip3] torchprofile==0.0.4
[pip3] torchvision==0.22.0a0+95f10a4e
[pip3] torchx==0.7.0
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by codingWizard-Nikhil**
Fixes #166166.

## Details
The TransformerEncoder's sparsity fast path was being incorrectly chosen when an `attn_mask` (named `mask`) was provided. This caused two problems:

1.  Inefficient performance and high VRAM usage on CUDA (the original issue).
2.  A `NotImplementedError` crash on MPS, as it tries to call a non-existent operator.

This PR adds a check for `mask is not None` to the `why_not_sparsity_fast_path` logic in `nn/modules/transformer.py`, ensuring the default path is used in this case. This fixes the issue on both CUDA and MPS.

## Test Plan
Added `test_transformer_encoder_fastpath_mask_check` which:
- Tests TransformerEncoder with both `mask` and `src_key_padding_mask` provided
- Verifies execution completes without errors on CUDA/MPS devices
- Runs with fastpath explicitly enabled to ensure the fix works

### CODE REVIEW

**Comment by codingWizard-Nikhil** [comment]
Hello, I see all CI checks are green on this PR. Please let me know if there's anything else needed from my side to help move this forward. Thanks!

**Comment by codingWizard-Nikhil** [comment]
@janeyx99 I submitted this PR around a month ago. Is this just part of the process or am I missing something?

**Review by mikaylagawarecki — COMMENTED** [review]
Thanks for the PR, could you add benchmarks to the PR description to illustrate the issue this fixes please

The initial issue mentions vram/performance but your unittest mentions a crash
